### DISCUSSION TITLE
Binding problem - bug or wrong code?

### DISCUSSION THREAD

**Discussion by PuerNoctis** [unoplatform/uno#9272] [Q&A]
Hi,

I have a slight problem with one of my bindings, and I don't know if there is a problem with my code, or if it may be a bug. The situation is the following:

I have a list of models that come from a backend and I want to display each item in a list using a custom control. For a minimum example, the model has just a single string property which I want to display in a text box and no view models are used. For this example, the list is being filled during the `Loaded` event of the page.

I will describe the exact problem further below. I am using Uno version 1.5.0 with Visual Studio 17.3.0 Preview 2.

First, here's all the code:
  - The model that holds the data of the model
  - The custom control representing a list item, used in a `DataTemplate` later on (XAML and code behind)
  - The MainPage using a `ListView` and binding the list of models (XAML and code behind)

### The model:
```csharp
using CommunityToolkit.Mvvm.ComponentModel;

namespace UnoTest.Models
{
    public class MyModel : ObservableObject
    {
        private string _someText;

        public string SomeText
        {
            get => _someText;
            set => SetProperty(ref _someText, value);
        }
    }
}
```

### Custum Control (XAML)
```xaml
<UserControl
    x:Class="UnoTest.Controls.MyControl"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:UnoTest.Controls"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"
    d:DesignHeight="300"
    d:DesignWidth="400">

    <TextBlock Text="{x:Bind Model.SomeText}" />
</UserControl>
```

### Custom Control (Code behind)
```csharp
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using UnoTest.Models;

namespace UnoTest.Controls
{
    public sealed partial class MyControl : UserControl
    {
        public static DependencyProperty ModelDependencyProperty = DependencyProperty.Register(
            nameof(Model), typeof(MyModel), typeof(MyControl), new PropertyMetadata(new MyModel { SomeText = "Initial Value" }));

        public MyModel Model
        {
            get => (MyModel)GetValue(ModelDependencyProperty);
            set => SetValue(ModelDependencyProperty, value);
        }

        public MyControl()
        {
            this.InitializeComponent();
        }
    }
}
```

### MainPage (XAML)
```xaml
<Page
    x:Class="UnoTest.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:UnoTest"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    xmlns:Controls="using:UnoTest.Controls"
    xmlns:Models="using:UnoTest.Models"
    mc:Ignorable="d"    
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"
    Loaded="OnLoaded">

    <StackPanel Orientation="Vertical" HorizontalAlignment="Stretch" VerticalAlignment="Stretch">
        <TextBlock Text="Here are some values:" Margin="20" FontSize="30" />

        <ListView ItemsSource="{x:Bind Models}">
            <ListView.ItemTemplate>
                <DataTemplate x:DataType="Models:MyModel">
                    <Controls:MyControl Model="{x:Bind}" />
                </DataTemplate>
            </ListView.ItemTemplate>
        </ListView>
    </StackPanel>
</Page>
```

### MainPage (Code behind)
``` csharp
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Controls.Primitives;
using Microsoft.UI.Xaml.Data;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Microsoft.UI.Xaml.Navigation;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using UnoTest.Models;
using Windows.Foundation;
using Windows.Foundation.Collections;

namespace UnoTest
{
    public sealed partial class MainPage : Page
    {
        public ObservableCollection<MyModel> Models { get; set; } = new ObservableCollection<MyModel>();

        public MainPage()
        {
            this.InitializeComponent();
        }

        public void OnLoaded(object sender, RoutedEventArgs e)
        {
            Models.Add(new MyModel { SomeText = "First" });
            Models.Add(new MyModel { SomeText = "Second" });
            Models.Add(new MyModel { SomeText = "Third" });
        }
    }
}
```

Now the thing is, when running this with the 'Windows' platform, I indeed get three entries in my list, but all with `Initial value` as per default of the dependency property. When running this with the 'Wasm' platform, it works perfectly fine and I get the `First`, `Second` and `Third` value respectively:

### Windows platform output
![image](https://user-images.githubusercontent.com/7012361/178954018-b436a455-9a85-4a66-99ac-45e78979cdc2.png)

### Wasm platform output
![image](https://user-images.githubusercontent.com/7012361/178954111-ecf6231e-ef34-49d0-af7a-b93e9c18f8ae.png)

When _not_ using a custom control but directly using a `TextBlock` in the `ListView` template, it works perfectly on Windows. Also, when defining a change callback in the dependency property, I get the correct `NewValue` when the value is set, so the DP _does_ receive a correct value.

Can anyone confirm, if there is something off with my bindings, or does the problem lie somewhere else?

Any help is very much appreciated, thanks!

**Comment by cesarchefinho** [comment]
Try Change the line

private string _someText;

To

private string _someText = string. Empty;

To prevent null value in property

Also

Change the line

<Controls:MyControl Model="{x:Bind}" />

To 

<Controls:MyControl/>

To prevent binding model to unknowvalue, 

Because The model is already assigned in your code behind cs file

**Reply by PuerNoctis** [reply]
Unfortunately, the suggestions didn't fix the issue :(

I am also not sure what you mean with the model being assigned in code behind? I only fill the observable list, the model DP still has to be bound to _something_ in the user control.

But I found some interesting things:

When using `{Binding}` in the text block of the control, the code works perfectly fine:

```xaml
<TextBlock Text="{Binding Model.SomeText, ElementName=self}" />
```

This lead me to believe, that there might be something awry with the code that `{x:Bind}` generates. So when stepping into said generated code during debug, I found that the method

```csharp
private void Update_Model(global::UnoTest.Models.MyModel obj, int phase)
```

only gets called with the initial model from the DP metadata. The generated code includes some methods to control the bindings, so I added the following change callback to my DP:

```csharp
public static DependencyProperty ModelDependencyProperty = DependencyProperty.Register(
    nameof(Model), typeof(MyModel), typeof(MyControl), new PropertyMetadata(new MyModel { SomeText = "Initial Value" }, (obj, value) =>
    {
        var ctrl = (MyControl)obj;
        ctrl.Bindings.Update();
    }));
```

And look here, the update method gets called with the correct values. The change callback does nothing but to update the bindings, which one would assume should be done by the framework when the value is set.

Overall I am still not sure if this is a bug with Uno generated `{x:Bind}` logic, or if there's something else wrong somewhere, but for the time being I have two workarounds that are not so bad.

I will keep track of the issue for a few days, and if there is no other input I will provide the above as an answer.

**Reply by jeromelaban** [reply]
Uno is behaving a bit differently when handling x:Bind (because of https://github.com/unoplatform/uno/issues/3519). Can you try adding `Mode=OneWay` to your binding, so that it gets updated when changed on Windows?
